tcms/core/ajax.py

```diff
@@ -235,27 +235,51 @@ def remove(self):
         tag_actions = TagActions(obj=obj, tag=q_tag)
         getattr(tag_actions, q_action)()
 
-    del q_tag, q_action
-
-    tags = obj.tag.all()
-    tags = tags.extra(select={
-        'num_plans':
-            'SELECT COUNT(*) FROM test_plan_tags '
-            'WHERE test_tags.tag_id = test_plan_tags.tag_id',
-        'num_cases':
-            'SELECT COUNT(*) FROM test_case_tags '
-            'WHERE test_tags.tag_id = test_case_tags.tag_id',
-        'num_runs':
-            'SELECT COUNT(*) FROM test_run_tags '
-            'WHERE test_tags.tag_id = test_run_tags.tag_id',
-    })
+    tags = obj.tag.all().order_by('pk')
+    test_plan_tags = TestPlanTag.objects.filter(
+        tag__in=tags).values('tag').annotate(num_plans=Count('tag')).order_by('tag')
+    test_case_tags = TestCaseTag.objects.filter(
+        tag__in=tags).values('tag').annotate(num_cases=Count('tag')).order_by('tag')
+    test_run_tags = TestRunTag.objects.filter(
+        tag__in=tags).values('tag').annotate(num_runs=Count('tag')).order_by('tag')
+
+    plan_counter = _TagCounter('num_plans', test_plan_tags)
+    case_counter = _TagCounter('num_cases', test_case_tags)
+    run_counter = _TagCounter('num_runs', test_run_tags)
+
+    for tag in tags:
+        tag.num_plans = plan_counter.calculate_tag_count(tag)
+        tag.num_cases = case_counter.calculate_tag_count(tag)
+        tag.num_runs = run_counter.calculate_tag_count(tag)
+
     context_data = {
         'tags': tags,
         'object': obj,
     }
     return render(request, template_name, context_data)
 
 
+class _TagCounter:
+    """ Used for counting the number of times a tags is assigned to TestRun/TestCase/TestPlan """
+
+    def __init__(self, key, test_tags):
+        self.key = key
+        self.test_tags = iter(test_tags)
+        self.counter = {'tag': 0}
+
+    def calculate_tag_count(self, tag):
+        if not self.counter['tag'] == tag.id:
+            try:
+                self.counter = self.test_tags.__next__()
+            except StopIteration:
+                return 0
+
+        if tag.pk == self.counter['tag']:
+            return self.counter[self.key]
+        else:
```

Comment:
Remove the else block and leave the return as the last return statement in the function. 

FYI there's a pylint checker that will catch this.